pytools/graph.py

```diff
@@ -520,5 +523,61 @@ def dfs(node: NodeT) -> None:
 
     return visited == graph.keys()
 
+# }}}
+
+
+# {{{ tag propagation graph processing
+
+def get_propagation_graph_from_constraints(
+        constraints: Iterable[Tuple[str, str]],
+    ) -> Mapping[str, FrozenSet[str]]:
```

Comment:
Is `str` really essential here? Could this be `GraphNodeT`, so that this could return `GraphT`?